### DISCUSSION TITLE
How to include total for cursor pagination?

### DISCUSSION THREAD

**Discussion by holtgrewe** [uriyyo/fastapi-pagination#960] [Q&A] [answered]
How can I enable the "total" field in the result page for sqlalchemy cursor pagination?

**Comment by uriyyo** [comment, marked as answer]
Hi @holtgrewe,

Currently, you need to define a custom page for this:
```py
from typing import Generic, TypeVar

from fastapi_pagination.bases import CursorRawParams
from fastapi_pagination.cursor import CursorPage, CursorParams

T = TypeVar("T")


class TotalCursorParams(CursorParams):
    def to_raw_params(self) -> CursorRawParams:
        params = super().to_raw_params()
        params.include_total = True

        return params


class TotalCursorPage(CursorPage[T], Generic[T]):
    __params_type__ = TotalCursorParams
```

**Reply by holtgrewe** [reply]
Thanks!

**Comment by zromick** [comment]
This issue should be a part of the documentation for Cursor if it's not already.
But also, is there a feature request out for implementing this? There should be one :))

Also, the provided solution didn't work for me.

**Reply by uriyyo** [reply]
Hi @zromick,

I updated docs, please let me know if it still not clear.

https://uriyyo-fastapi-pagination.netlify.app/tutorials/cursor-pagination/

**Reply by zromick** [reply]
```
	"total": null,
	"current_page": "Pg%3D%3D",
	"current_page_backwards": "PGR0OjIwMjQtMDYtMDYgMTQ6MzQ6NTkuNjA0Mjg2",
	"previous_page": null,
	"next_page": "PmR0OjIwMjQtMDYtMDYgMTQ6MzU6MDAuMzczMTc4"
```

Thanks, uriyyo! I'm still getting a null total after adding that code.

Also, there really ought to be one more example that uses CursorParams. It was difficult to figure out that I needed that in order to get things fully working:
```
        params = CursorParams(cursor=cursor, size=size)
        return paginate(session, select_statement, params)
```

Lastly, I think the cursor examples located here could also include CursorParams and total:
https://github.com/uriyyo/fastapi-pagination/blob/main/examples/cursor/app.py

I'd also suggest replacing/removing .with_custom_options since you've deprecated it.
```
    @deprecated(
        "`with_custom_options` method is deprecated, please use "
        "`fastapi_pagination.customization.CustomizePage` instead. "
        "This method will be removed in the next major release (0.13.0)."
    )
```

The reason for all of these recommendations is because it took me several hours to figure out how to use cursor :))
In sum, I'd rather have a completely documented version 12 over a finished version 13.
I do know that you're the only active contributor here, though, so no rush haha. Keep up the great work!

**Reply by uriyyo** [reply]
Yup, documentation is big issue, I'm not good at it. I'm trying to find somebody who will help me with it, but no progress for now(

I updated cursor example, could you please check it out? - https://github.com/uriyyo/fastapi-pagination/blob/main/examples/cursor/app.py

> Lastly, I think the cursor examples located here could also include CursorParams and total:

Hmmm, actually you don't need to specify them explicitly. Could you please show example how you define your routes?

**Reply by zromick** [reply]
Here's the gist:
```
@router.get("/items", response_model=CursorPage[Item])
def get_items(
    cursor: Optional[str] = None,
    size: int = 10,
    session: Session = Depends(get_db),
):
    return crud.get_items(session, cursor, size)
  ```
    
   in `crud.get_items`:
```
def get_items(
        self,
        session: Session,
        cursor: Optional[str] = None,
        size: int = 10,
    ):
        select_statement = select(models.Item).order_by(models.Item.item_created_time.desc())
        params = CursorParams(cursor=cursor, size=size)
        items = paginate(session, select_statement, params)
        # items.total = not working

        return items
```

**Reply by uriyyo** [reply]
Could you please show what `CursorPage` looks like?

It should be like this:
https://github.com/uriyyo/fastapi-pagination/blob/f310fd8d81cb1fa4ac9db42bb51e0e5e75f7fb9a/examples/cursor/app.py#L28-L32

Also, don't forget to add `add_pagination` call with your app.

**Comment by zromick** [comment]
Yeah I didn't include `add_pagination`, but its there and working 👍 

My CursorPage isn't customized right now, it's just imported `from fastapi_pagination.cursor import CursorPage`

I'm not sure if CustomizedPage would be a good replacement, since don't I need to set params each time via the cursor and size params (as chosen by the user during the call)?

**Reply by uriyyo** [reply]
> My CursorPage isn't customized right now, it's just imported from fastapi_pagination.cursor import CursorPage

You need to customize your page, without it it will not include `total` to page 😢 

>  since don't I need to set params each time via the cursor and size params (as chosen by the user during the call)?

If you add `add_pagination` then there will be no need to pass size and cursor values, it will be done automatically.

**Reply by zromick** [reply]
Hey, it works! I think manually setting `CursorParams` was overriding the `CustomizedPage` you provided.

Quick question, how might I add a custom field to the response? Is there documentation for that somewhere?

```
{
	"items": [
		{
			...
		}
	],
	"total": 10,
	"current_page": "",
	"current_page_backwards": "",
	"previous_page": "",
	"next_page": ""
        "custom_field": True
}
```

**Comment by h-srinivasan** [comment]
I'm unable to get this to work with cassandra, could someone please guide me or show an example of how to implement it? Thanks!
